eventuals/interrupt.h

```diff
@@ -63,22 +129,59 @@ class Interrupt final {
       return false;
     }
 
+    Handler::State state = handler->state_.state();
+    CHECK_EQ(state, Handler::State::UNINSTALLED);
+
     bool installed = handler_.compare_exchange_weak(
         stored,
         handler,
         std::memory_order_release,
         std::memory_order_relaxed);
 
-    // 'stored' might be a 'nullptr' only in case of race betweer
-    // 'load' and 'compare_exchange_weak'.
+    // 'stored' is 'nullptr' only when race between 'load' and
+    // 'compare_exchange_weak' occurs, so 'stored' is not stored
+    // in the 'handler_'.
     CHECK(installed || stored == nullptr);
+
+    CHECK(handler->state_.Update(state, Handler::State::INSTALLED));
+
+    if (stored != nullptr) {
+      Handler::State state = stored->state_.state();
+      while (!stored->state_.Update(state, Handler::State::UNINSTALLED)) {}
+    }
+
     return installed;
   }
 
+  [[nodiscard]] bool Uninstall(Handler* handler) {
+    // If 'handler' is current 'handler_' then the interrupt *must not*
+    // have been triggered so we should set it back to 'placeholder_handler_'
+    // instead of nullptr. If we successfully did the compare-and-swap then
+    // we can update our state to UNINSTALLED.
+    //
+    // Otherwise someone else must have done the compare-and-swap
+    // and they are responsible for setting us to UNINSTALLED.
+    if (handler_.compare_exchange_weak(
+            handler,
+            &placeholder_handler_,
+            std::memory_order_release,
+            std::memory_order_relaxed)) {
+      Handler::State state = handler->state_.state();
+      CHECK_EQ(state, Handler::State::UNINSTALLING);
+      CHECK(handler->state_.Update(state, Handler::State::UNINSTALLED));
+      return true;
+    }
+    return false;
+  }
+
   void Trigger() {
     // NOTE: nullptr signifies that the interrupt has been triggered.
     Handler* handler = handler_.exchange(nullptr);
-    if (handler != nullptr) {
+    if (handler != nullptr && handler != &placeholder_handler_) {
+      Handler::State state = handler->state_.state();
+      // To be sure that callback won't be invoked while destructing.
+      CHECK_EQ(state, Handler::State::INSTALLED);
+      CHECK(handler->state_.Update(state, Handler::State::UNINSTALLED));
```

Comment:
```suggestion
      // Need to uninstall the handler so that in the event invoking the callback causes the handler to get destructed it won't try and uninstall itself!
      CHECK_EQ(state, Handler::State::INSTALLED);
      CHECK(handler->state_.Update(state, Handler::State::UNINSTALLED));
```